### ISSUE TITLE
Server does not disconnect invalid client

### ISSUE THREAD

**Issue by jacobson** [ratchetphp/Ratchet#334]
While using WsServer, server does not disconnect basic/plain telnet connections.

Use http://socketo.me/docs/hello-world and telnet to server, whatever you do, you will remain connected.

**Comment by cboden** [comment]
Hit ctrl+] then type "quite" and hit return

**Comment by jacobson** [comment]
You are kidding, right? (:

I mean: according to RFC, websocket protocol requires disconnection of the client if no handshake is possible. But not just for RFC, I want to detect, reject or disconnect those connections from server so they wont abuse the system or consume resources.

**Comment by cboden** [comment]
I think misunderstood your initial question. Can you elaborate on your issue and steps to reproduce? 

If a valid handshake is not established the server will disconnect the client. By default if the client sends over 4kb of data without it establishing a handshake the server will disconnect the client.

**Comment by jacobson** [comment]
$server = IoServer::factory(
    new HttpServer(
        new WsServer(
            new Chat()
        )
    ),
    8080
);

Without any specific reason I connect to Ratchet server with telnet localhost 8080. For sure it does nothing as the server expects WS protocol.

At this point, it is obivous that telnet is an invalid client for my app. So I don't want it to hang around and user my server's resources.

What can I do for this purpose? Is there an option or parameter I can specify to disconnect this connection or can I detect it easily with serverside code and disconnect it from there with some kind of fingerprint/agent or with a timeout? It seems like I have to apply a manual method to accomplish this and return HTTP 400 status code.

**Comment by cboden** [comment]
WebSockets are just a protocol on top of TCP sockets. Telnet is a TCP socket client. The connection telnet is establishing is the same that a web browser first establishes before initializing a WebSocket handshake. Until data is transmitted a server has no idea what the client's intentions are. 

The only thing I can think of to do differently is disconnect the client after a certain period of time if it hasn't initialized a WebSocket handshake. You can do this with the React `EventLoop` you pass to `IoServer` using `setTimer`.

**Comment by jacobson** [comment]
Fair enough. I will look into that. Meanwhile, is the following expected behaviour? -->

- Telnet localhost 8080
- Hit enter 2 times
- Whatever you send, you won't get disconnected by the server.

Exit the telnet session, and start a new one:
- Telnet localhost 8080
- Type anything (asdassdfsfsdf)
- Hit enter 2 times.
- Server will disconnect you with "Bad Request". Which is correct.

Is this the expected behaviour? I think the validation part no longer works if it receives "\n\n" thinking that headers are already received; but it doesn't check if the headers are even not empty or valid.

If this is the expected behaviour, I will also add another listener manually to validate the input after "\n\n" if there is nothing before "\n\n".

**Comment by cboden** [comment]
I believe this is expected behaviour. I telnet'd into google.com and tried this, it seems parsers begin looking for headers when non-empty values are sent.

**Comment by jacobson** [comment]
After your message, I tried it with google and even though it is similar it is not exactly the same:

- Telnet google.com 80
- Hit enter 2 times
- Type asdasasdas and hit enter
- You will be disconnected with Bad Request error.

Repeat the same with ratchet server, you won't get disconnected if you hit enter 2 times before entering anything.

**Comment by PaulRotmann** [comment]
@CharlieReece Thanks for reporting and for your patience, I know it's been a while! :)

We've merged #1110 which fixes rejecting invalid HTTP requests starting with newlines. This is part of a larger effort reviving Ratchet as discussed in #1054, solving many other issues along the way.

If you'd like to support us on this journey, consider [sponsoring](https://github.com/sponsors/reactphp) the project – sponsorships really do make a difference. ❤️
